### PR TITLE
[21313] Type propagation policy

### PR DESCRIPTION
**Pull request by EduPonz**
## Description

This PR adds a property policy to the `DomainParticipant` to control how types are propagated by EDP and Type Lookup Service.
The property name is `fastdds.type_propagation`
The property can take 4 values:

| Value | Register Type Object | Send TypeInformation on EDP | Receive TypeInformation on EDP | Type Lookup replies |
|-|-|-|-|-|
| `disabled` | NO | NO | IGNORED | NO |
| `enabled` | BOTH | BOTH | PROCESSED | COMPLETE & COMPLETE_TO_MINIMAL |
| `minimal_bandwidth` | MINIMAL | MINIMAL | Only MINIMAL PROCESSED | MINIMAL |
| `registration_only` | BOTH | NO | IGNORED | NO |

Related PRs:

* IDL Parser: eProsima/IDL-Parser#152
* Fast DDS Gen: eProsima/Fast-DDS-Gen#380
* Fast DDS Docs: eProsima/Fast-DDS-docs#880
* Shapes Demo: eProsima/ShapesDemo#167
* Shapes Demo Docs: eProsima/Shapes-Demo-Docs#129

## Contributor Checklist

- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [x] Any new configuration API has an equivalent XML API (with the corresponding XSD extension) 
- ❌ Changes are backport compatible: they do **NOT** break ABI nor change library core behavior. 
- ❌ Changes are API compatible. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- [x] New feature has been documented/Current behavior is correctly described in the documentation. 
    
- _N/A_: Applicable backports have been included in the description.

## Reviewer Checklist

- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- _N/A_: If this is a critical bug fix, backports to the critical-only supported branches have been requested.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/fastdds/builtin/type_lookup_service/TypeLookupManager.cpp [hunk 1]
```diff
@@ -118,6 +118,8 @@ bool TypeLookupManager::init(
         protocols->mp_participantImpl->getRTPSParticipantAttributes().allocation.locators.max_unicast_locators,
         protocols->mp_participantImpl->getRTPSParticipantAttributes().allocation.locators.max_multicast_locators);
 
+    auto properties = participant_->getAttributes().properties;
+    type_propagation_ = utils::to_type_propagation(properties);
```
Comment (MiguelCompany) [inline src/cpp/fastdds/builtin/type_lookup_service/TypeLookupManager.cpp:122]: This pattern repeats a lot. We could add a method in `RTPSParticipantImpl` to get the type propagation.

File: src/cpp/fastdds/publisher/DataWriterImpl.cpp [hunk 2]
```diff
@@ -1685,12 +1686,50 @@ fastdds::TopicAttributes DataWriterImpl::get_topic_attributes(
...
+                    // Use minimal type information only
+                    topic_att.type_information.type_information.minimal() = type_info.minimal();
+                    break;
+                }
+                case TypePropagation::TYPEPROPAGATION_REGISTRATION_ONLY:
+                {
+                    // Use both complete and minimal type information
+                    topic_att.type_information.type_information = type_info;
+                    break;
+                }
```
Comment (MiguelCompany) [inline src/cpp/fastdds/publisher/DataWriterImpl.cpp:1726]: The conditional on `should_assign_type_information` makes this case impossible to reach

```suggestion
```

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 3]
```diff
@@ -1778,12 +1779,50 @@ fastdds::TopicAttributes DataReaderImpl::topic_attributes() const
...
+                    // Use minimal type information only
+                    topic_att.type_information.type_information.minimal() = type_info.minimal();
+                    break;
+                }
+                case TypePropagation::TYPEPROPAGATION_REGISTRATION_ONLY:
+                {
+                    // Use both complete and minimal type information
+                    topic_att.type_information.type_information = type_info;
+                    break;
+                }
```
Comment (MiguelCompany) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:1819]: The conditional on `should_assign_type_information` makes this case impossible to reach

```suggestion
```

File: test/examples/xtypes_complete.compose.yml [hunk 4]
```diff
@@ -0,0 +1,89 @@
+# Copyright 2024 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
Comment (MiguelCompany) [inline test/examples/xtypes_complete.compose.yml:1]: Should we also test minimal <-> complete interactions?

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: examples/cpp/xtypes/SubscriberApp.cpp [hunk 5]
```diff
@@ -197,49 +227,56 @@ void SubscriberApp::on_data_writer_discovery(
...
         // Get remote type information and use it to retrieve the type object
-        auto type_info = info.type_information.type_information;
-        auto type_id = type_info.complete().typeid_with_size().type_id();
-
-        if (RETCODE_OK != DomainParticipantFactory::get_instance()->type_object_registry().get_type_object(
-                    type_id,
-                    remote_type_object_))
+        remote_type_information_ = info.type_information.type_information;
+        auto type_id_complete = remote_type_information_.complete().typeid_with_size().type_id();
+        auto type_id_minimal = remote_type_information_.minimal().typeid_with_size().type_id();
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply examples/cpp/xtypes/SubscriberApp.cpp:238]: I don't think so. Looking at those functions, none of them throws and they all return references

File: src/cpp/fastdds/domain/DomainParticipantImpl.cpp [hunk 6]
```diff
@@ -1700,12 +1708,30 @@ bool DomainParticipantImpl::set_qos(
...
+    {
+        // Check participant's type propagation policy
+        using utils::to_type_propagation;
+        using utils::TypePropagation;
+
+        auto type_propagation = to_type_propagation(qos.properties());
+
+        if (TypePropagation::TYPEPROPAGATION_UNKNOWN == type_propagation)
+        {
+            EPROSIMA_LOG_ERROR(RTPS_QOS_CHECK, "Invalid value for fastdds.type_propagation property");
```
Comment (MiguelCompany) [inline src/cpp/fastdds/domain/DomainParticipantImpl.cpp:1729]:
```suggestion
            EPROSIMA_LOG_ERROR(RTPS_QOS_CHECK, "Invalid value for " << parameter_policy_type_propagation << " property");
```

File: src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp [hunk 7]
```diff
@@ -503,10 +503,12 @@ void PDPServer::initializeParticipantProxyData(
...
 
-    if (getRTPSParticipant()->getAttributes().builtin.discovery_config.discoveryProtocol !=
+    auto discovery_config = getRTPSParticipant()->get_attributes().builtin.discovery_config;
+
+    if (discovery_config.discoveryProtocol !=
             DiscoveryProtocol::SERVER
             &&
-            getRTPSParticipant()->getAttributes().builtin.discovery_config.discoveryProtocol !=
+            discovery_config.discoveryProtocol !=
             DiscoveryProtocol::BACKUP)
```
Comment (MiguelCompany) [inline src/cpp/rtps/builtin/discovery/participant/PDPServer.cpp:512]:
```suggestion
    if (DiscoveryProtocol::SERVER != discovery_config.discoveryProtocol &&
            DiscoveryProtocol::BACKUP != discovery_config.discoveryProtocol)
```

File: test/mock/rtps/RTPSParticipantImpl/rtps/participant/RTPSParticipantImpl.h [hunk 8]
```diff
@@ -309,12 +294,7 @@ class RTPSParticipantImpl
         return 65536;
     }
 
-    const RTPSParticipantAttributes& getRTPSParticipantAttributes() const
-    {
-        return attr_;
-    }
-
-    RTPSParticipantAttributes& getAttributes()
+    RTPSParticipantAttributes get_attributes() const
```
Comment (MiguelCompany) [inline test/mock/rtps/RTPSParticipantImpl/rtps/participant/RTPSParticipantImpl.h:297]:
```suggestion
    const RTPSParticipantAttributes& get_attributes() const
```

File: test/examples/xtypes_complete.compose.yml (hunk 4 above)
Comment (MiguelCompany) [reply test/examples/xtypes_complete.compose.yml:1]: We will leave this for a follow-up

**Review by MiguelCompany — DISMISSED** [review]
LGTM with green CI

**Review by MiguelCompany — APPROVED** [review]

**Review by MiguelCompany — APPROVED** [review]

**Pull request merged**